velox/functions/prestosql/tests/ArrayShuffleTest.cpp

```diff
@@ -0,0 +1,325 @@
+    } else {
+      out << "NULL";
+    }
+    out << ", ";
+  });
+  out << "]";
+  return out.str();
+}
+
+template <typename T>
+std::string printArray(
+    const std::vector<std::optional<std::vector<std::optional<T>>>>& input) {
+  std::stringstream out;
+  out << "[";
+  std::for_each(input.begin(), input.end(), [&out](const auto& e) {
+    if (e.has_value()) {
+      out << printArray(e.value());
+    } else {
+      out << "NULL";
+    }
+    out << ", ";
+  });
+  out << "]";
+  return out.str();
+}
+
+template <typename T>
+std::string printArray(const std::vector<T>& input) {
+  std::stringstream out;
+  out << "[";
+  std::for_each(
+      input.begin(), input.end(), [&out](const auto& e) { out << e << ", "; });
+  out << "]";
+  return out.str();
+}
+
+class ArrayShuffleTest : public FunctionBaseTest {
+ protected:
+  constexpr static const int32_t kNumOfEvalTimes = 100;
+  constexpr static const int32_t kNums = 10;
+  constexpr static const double kUniquenessRate = .7;
+
+  void testTwoExprsEquals(
+      const std::string& expression1,
+      const std::string& expression2,
+      const std::vector<VectorPtr>& input) {
+    auto result1 = evaluate<ArrayVector>(expression1, makeRowVector(input));
+    auto result2 = evaluate<ArrayVector>(expression2, makeRowVector(input));
+    assertEqualVectors(result1, result2);
+  }
+
+  template <typename T, typename Cont>
+  void decodeElementsAndTestIsAPermutation(
+      const BaseVector& actualVector,
+      const std::vector<std::vector<std::optional<T>>>& expected) {
+    DecodedVector decoded;
+    exec::VectorReader<Cont> reader(decode(decoded, actualVector));
+
+    for (auto i = 0; i < actualVector.size(); i++) {
+      ASSERT_TRUE(reader.isSet(i))
+          << "Actual array at " << i
+          << " is a null, which we don't need to test here";
+      auto actualArray = reader[i].materialize();
+      // 3. assert the two vectors containing the same elements (ignoring order)
+      LOG(WARNING) << "Actual array " << printArray(actualArray) << " at " << i
+                   << " vs expected array " << printArray(expected[i]);
+      ASSERT_TRUE(std::is_permutation(
+          actualArray.begin(), actualArray.end(), expected[i].begin()))
+          << "Actual array " << printArray(actualArray) << " at " << i
+          << " must produce a permutation of the expected array "
+          << printArray(expected[i]);
+    }
+  }
+
+  template <typename T, typename Cont = Array<T>>
+  void testIsPermutation(
+      const std::vector<std::vector<std::optional<T>>>& expected,
+      const std::string& expression,
+      const std::vector<VectorPtr>& input) {
+    auto arrayVectorResult =
+        evaluate<ArrayVector>(expression, makeRowVector(input));
+    decodeElementsAndTestIsAPermutation<T, Cont>(
+        *arrayVectorResult.get(), expected);
+  }
+
+  /// To test a shuffle, we will do the following steps:
+  /// 1. run the test for T times with range-based input {0, 1, ..., n-1},
```

Comment:
"evaluate shuffle T times on sequential input [0, 1, 2,... n-1]; verify that X% of the results are unique."